From 72fbdb2eb5589cf6d84899034d04b01722a64718 Mon Sep 17 00:00:00 2001
From: Yash Katariya <yashkatariya@google.com>
Date: Tue, 19 Dec 2023 16:30:48 -0800
Subject: [PATCH] Expose `shard_alike` via jax.experimental. The API is `x, y =
 shard_like(x, y)`.

The guarantee provided by this API is that the sharding of `x` and `y` will be the same! What the sharding will be is decided by GSPMD.

The flow of sharding is bidirectional i.e. SPMD will choose what the sharding should be of `x` and `y` depending on it's propagation algorithm. It might end up being that the sharding chosen is not of `x` and `y` but something better. At the end of propagation `x` and `y` will be sharded alike.

The API can be made variadic in the future i.e. `*args = shard_alike(*args)` depending on use cases.

Fixes: https://github.com/google/jax/issues/15600
PiperOrigin-RevId: 592375936
---
 jax/BUILD                         |   1 +
 jax/_src/interpreters/mlir.py     |   6 +-
 jax/_src/shard_alike.py           |  97 ++++++++++++
 jax/experimental/export/export.py |   3 +-
 jax/experimental/jax2tf/jax2tf.py |   1 +
 jax/experimental/shard_alike.py   |  17 ++
 tests/BUILD                       |   8 +
 tests/shard_alike_test.py         | 248 ++++++++++++++++++++++++++++++
 8 files changed, 378 insertions(+), 3 deletions(-)
 create mode 100644 jax/_src/shard_alike.py
 create mode 100644 jax/experimental/shard_alike.py
 create mode 100644 tests/shard_alike_test.py

diff --git a/jax/BUILD b/jax/BUILD
index c854f7b1cc32..02deeb543e46 100644
--- a/jax/BUILD
+++ b/jax/BUILD
@@ -209,6 +209,7 @@ py_library_providing_imports_info(
         "_src/prng.py",
         "_src/public_test_util.py",
         "_src/random.py",
+        "_src/shard_alike.py",
         "_src/stages.py",
     ] + glob(
         [
diff --git a/jax/_src/interpreters/mlir.py b/jax/_src/interpreters/mlir.py
index e1eefbf315af..5022fa5a59f6 100644
--- a/jax/_src/interpreters/mlir.py
+++ b/jax/_src/interpreters/mlir.py
@@ -2062,7 +2062,8 @@ def _wrap_with_spmd_op(name: str,
                        x: ir.Value,
                        aval_out: core.AbstractValue,
                        sharding_proto: xc.OpSharding,
-                       unspecified_dims: set[int] | None = None):
+                       unspecified_dims: set[int] | None = None,
+                       has_side_effect: bool = False):
   # unspecified_dims indicate dimensions whose shardings are not specified and
   # XLA sharding propagation can change them.
   if unspecified_dims:
@@ -2080,7 +2081,8 @@ def _wrap_with_spmd_op(name: str,
   op = custom_call(name, result_types=[result_type], operands=[x],
                    backend_config=backend_config,
                    api_version=1,
-                   result_shapes=result_shapes)
+                   result_shapes=result_shapes,
+                   has_side_effect=has_side_effect)
   set_sharding(op, sharding_proto)
   return op.result
 
diff --git a/jax/_src/shard_alike.py b/jax/_src/shard_alike.py
new file mode 100644
index 000000000000..951b535fc4b7
--- /dev/null
+++ b/jax/_src/shard_alike.py
@@ -0,0 +1,97 @@
+# Copyright 2023 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import partial
+import itertools
+
+from jax._src import core
+from jax._src.interpreters import ad
+from jax._src.interpreters import mlir
+from jax._src.dispatch import apply_primitive
+from jax._src.tree_util import tree_flatten, tree_unflatten
+from jax._src.interpreters import batching
+from jax._src.util import safe_zip
+from jax._src.lib import xla_client as xc
+from jax._src.api_util import shaped_abstractify
+from jax._src.lib.mlir import ir
+
+_next_shard_group_id = itertools.count()
+
+def shard_alike(x, y):
+  """Shards x and y alike."""
+  x_flat, x_tree = tree_flatten(x)
+  y_flat, y_tree = tree_flatten(y)
+
+  if x_tree != y_tree:
+    raise ValueError('Trees should be equal. '
+                     f'Got x_tree: {x_tree}, y_tree: {y_tree}')
+
+  for x_, y_ in safe_zip(x_flat, y_flat):
+    x_aval = shaped_abstractify(x_)
+    y_aval = shaped_abstractify(y_)
+    if x_aval.shape != y_aval.shape:
+      raise ValueError(
+          'The leaves shapes of `x` and `y` should match. Got `x` leaf shape:'
+          f' {x_aval.shape} and `y` leaf shape: {y_aval.shape}. File an issue at'
+          ' https://github.com/google/jax/issues if you want this feature.')
+
+  outs = [shard_alike_p.bind(x_, y_) for x_, y_ in safe_zip(x_flat, y_flat)]
+  x_out_flat, y_out_flat = zip(*outs)
+  return tree_unflatten(x_tree, x_out_flat), tree_unflatten(y_tree, y_out_flat)
+
+
+shard_alike_p = core.Primitive('shard_alike')
+shard_alike_p.multiple_results = True
+shard_alike_p.def_impl(partial(apply_primitive, shard_alike_p))
+shard_alike_p.def_abstract_eval(lambda x, y: (x, y))
+
+def shard_alike_transpose(ct, **kwargs):
+  x_ct, y_ct = ct
+  if type(x_ct) is ad.Zero or type(y_ct) is ad.Zero:
+    return x_ct, y_ct
+  else:
+    return shard_alike(x_ct, y_ct)
+ad.deflinear(shard_alike_p, shard_alike_transpose)
+
+def _shard_alike_batcher(vals_in, dims_in):
+  vals_out = shard_alike_p.bind(*vals_in)
+  return vals_out, dims_in
+batching.primitive_batchers[shard_alike_p] = _shard_alike_batcher
+
+
+def _group_shard(
+    ctx,
+    x: ir.Value,
+    y: ir.Value,
+    x_aval_out: core.AbstractValue,
+    y_aval_out: core.AbstractValue,
+) -> tuple[ir.Value, ir.Value]:
+  shard_group_id = next(_next_shard_group_id)
+
+  unknown_op_sharding = xc.OpSharding()
+  unknown_op_sharding.type = xc.OpSharding.Type.UNKNOWN
+  unknown_op_sharding.is_shard_group = True
+  unknown_op_sharding.shard_group_id = shard_group_id
+  unknown_op_sharding.shard_group_type = xc.OpSharding.ShardGroupType.AS
+
+  x = mlir.wrap_with_sharding_op(ctx, x, x_aval_out, unknown_op_sharding,
+                                 has_side_effect=True)
+  y = mlir.wrap_with_sharding_op(ctx, y, y_aval_out, unknown_op_sharding,
+                                 has_side_effect=True)
+  return x, y
+
+
+def shard_alike_lowering(ctx, x, y):
+  return _group_shard(ctx, x, y, *ctx.avals_out)
+mlir.register_lowering(shard_alike_p, shard_alike_lowering)
diff --git a/jax/experimental/export/export.py b/jax/experimental/export/export.py
index 1508a6f738a2..03fffc9eeccd 100644
--- a/jax/experimental/export/export.py
+++ b/jax/experimental/export/export.py
@@ -880,6 +880,7 @@ def _check_lowering(lowering) -> None:
     "shape_assertion",  # Used by shape_poly to evaluate assertions
 }
 
+check_sharding_pattern = re.compile(r"^({replicated}|{unknown shard_as.*}|"")$")
 
 def _check_module(mod: ir.Module, *,
                   allow_non_replicated_sharding: bool,
@@ -910,7 +911,7 @@ def check_sharding(op: ir.Operation, loc: ir.Location):
       except KeyError:
         pass
       else:
-        if ir.StringAttr(sharding).value not in ["{replicated}", ""]:
+        if not re.match(check_sharding_pattern, ir.StringAttr(sharding).value):
           raise ValueError(
               "Lowered function does not have a top-level pjit but it has"
               f" non-replicated sharding annotations, e.g., {op} at {loc}.\nSee"
diff --git a/jax/experimental/jax2tf/jax2tf.py b/jax/experimental/jax2tf/jax2tf.py
index 72502d153c7b..189d94c8cf6f 100644
--- a/jax/experimental/jax2tf/jax2tf.py
+++ b/jax/experimental/jax2tf/jax2tf.py
@@ -1508,6 +1508,7 @@ def _unexpected_primitive(p: core.Primitive, *args, **kwargs):
     "platform_index",
     "assert_consumed_value",
     "consume",
+    "shard_alike",
 ]
 
 tf_impl[unconsumed_copy_p] = lambda x: x
diff --git a/jax/experimental/shard_alike.py b/jax/experimental/shard_alike.py
new file mode 100644
index 000000000000..bc18e7f5ad61
--- /dev/null
+++ b/jax/experimental/shard_alike.py
@@ -0,0 +1,17 @@
+# Copyright 2023 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the ific language governing permissions and
+# limitations under the License.
+
+from jax._src.shard_alike import (
+    shard_alike as shard_alike
+)
diff --git a/tests/BUILD b/tests/BUILD
index 1cd3c0d6458e..8038bc57cc6c 100644
--- a/tests/BUILD
+++ b/tests/BUILD
@@ -262,6 +262,14 @@ jax_test(
     tags = ["multiaccelerator"],
 )
 
+jax_test(
+    name = "shard_alike_test",
+    srcs = ["shard_alike_test.py"],
+    deps = [
+        "//jax:experimental",
+    ],
+)
+
 jax_test(
     name = "pgle_test",
     srcs = ["pgle_test.py"],
diff --git a/tests/shard_alike_test.py b/tests/shard_alike_test.py
new file mode 100644
index 000000000000..7578524a1e55
--- /dev/null
+++ b/tests/shard_alike_test.py
@@ -0,0 +1,248 @@
+# Copyright 2021 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from absl.testing import absltest
+from jax._src import xla_bridge
+from jax._src import test_util as jtu
+from jax.sharding import NamedSharding, PartitionSpec as P
+from jax.experimental.shard_alike import shard_alike
+from jax.experimental.shard_map import shard_map
+from jax._src.lib import xla_extension_version
+
+from jax import config
+config.parse_flags_with_absl()
+
+prev_xla_flags = None
+
+
+def setUpModule():
+  global prev_xla_flags
+  prev_xla_flags = os.getenv("XLA_FLAGS")
+  flags_str = prev_xla_flags or ""
+  # Don't override user-specified device count, or other XLA flags.
+  if "xla_force_host_platform_device_count" not in flags_str:
+    os.environ["XLA_FLAGS"] = (flags_str +
+                               " --xla_force_host_platform_device_count=8")
+  # Clear any cached backends so new CPU backend will pick up the env var.
+  xla_bridge.get_backend.cache_clear()
+
+def tearDownModule():
+  if prev_xla_flags is None:
+    del os.environ["XLA_FLAGS"]
+  else:
+    os.environ["XLA_FLAGS"] = prev_xla_flags
+  xla_bridge.get_backend.cache_clear()
+
+
+class ShardAlikeTest(jtu.JaxTestCase):
+
+  def setUp(self):
+    super().setUp()
+    if xla_extension_version < 227:
+      self.skipTest('Requires xla_extension_version >= 227')
+
+  def test_basic(self):
+    mesh = jtu.create_global_mesh((2, 2), ('x', 'y'))
+    np_inp = np.arange(16).reshape(8, 2)
+    s = NamedSharding(mesh, P('x', 'y'))
+    inp = jax.device_put(np_inp, s)
+
+    @jax.jit
+    def f(x):
+      y = x * x
+      z = y * 2
+      _, z = shard_alike(x, z)
+      return z * 2
+
+    out = f(inp)
+    self.assertEqual(out.sharding, s)
+    self.assertArraysEqual(out, np_inp * np_inp * 4)
+
+  def test_output_sharded_alike_input(self):
+    mesh = jtu.create_global_mesh((2, 1), ('x', 'y'))
+    np_inp = np.arange(16).reshape(8, 2)
+    s = NamedSharding(mesh, P('x', 'y'))
+    inp = jax.device_put(np_inp, s)
+
+    @jax.jit
+    def f(x):
+      y = x * 2
+      return shard_alike(x, y)[1]
+
+    out = f(inp)
+    self.assertEqual(out.sharding, s)
+    self.assertArraysEqual(out, np_inp * 2)
+
+  def test_arange_shard_alike_jit(self):
+    mesh = jtu.create_global_mesh((2, 1), ('x', 'y'))
+    np_inp = np.arange(16).reshape(8, 2)
+    s = NamedSharding(mesh, P('x', 'y'))
+    inp = jax.device_put(np_inp, s)
+
+    @jax.jit
+    def f(x):
+      y = jnp.arange(16).reshape(8, 2)
+      return shard_alike(x, y)[1]
+
+    out = f(inp)
+    self.assertEqual(out.sharding, s)
+    self.assertArraysEqual(out, np_inp)
+
+  def test_different_shapes(self):
+    mesh = jtu.create_global_mesh((2, 1), ('x', 'y'))
+    np_inp = np.arange(16).reshape(8, 2)
+    s = NamedSharding(mesh, P('x',))
+    inp = jax.device_put(np_inp, s)
+
+    @jax.jit
+    def f(x):
+      y = x @ x.T
+      return shard_alike(x, y)[1]
+
+    with self.assertRaisesRegex(
+        ValueError, 'The leaves shapes of `x` and `y` should match'):
+      f(inp)
+
+  def test_double_shard_alike(self):
+    mesh = jtu.create_global_mesh((2, 2), ('x', 'y'))
+    np_inp = np.arange(16).reshape(8, 2)
+    s = NamedSharding(mesh, P('x', 'y'))
+    inp = jax.device_put(np_inp, s)
+
+    @jax.jit
+    def f(x):
+      y = x * 2
+      _, y = shard_alike(x, y)
+      z = y @ y.T
+      a = jnp.arange(64).reshape(8, 8)
+      return shard_alike(z, a)
+
+    out1, out2 = f(inp)
+    self.assertEqual(out1.sharding, NamedSharding(mesh, P('x')))
+    self.assertEqual(out2.sharding, NamedSharding(mesh, P('x')))
+
+  def test_shard_like_eager(self):
+    mesh = jtu.create_global_mesh((4, 1), ('x', 'y'))
+    np_inp = np.arange(16).reshape(8, 2)
+    s = NamedSharding(mesh, P('x', 'y'))
+    inp = jax.device_put(np_inp, s)
+
+    def f(x):
+      y = jnp.arange(16).reshape(8, 2)
+      return shard_alike(x, y)[1]
+
+    out = f(inp)
+    self.assertEqual(out.sharding, s)
+    self.assertArraysEqual(out, np_inp)
+
+  def test_shard_map(self):
+    mesh = jtu.create_global_mesh((4, 2), ('x', 'y'))
+    np_inp = np.arange(16).reshape(8, 2)
+    s = NamedSharding(mesh, P('x', 'y'))
+    inp = jax.device_put(np_inp, s)
+
+    def g(x):
+      return jax.lax.psum(x, 'x')
+
+    @jax.jit
+    def f(x):
+      y = x @ x.T
+      s_out = shard_map(g, mesh, in_specs=P('x', 'y'),
+                        out_specs=P(None, 'y'))(y)
+      z = s_out.T @ s_out
+      return shard_alike(y, z)
+
+    out1, out2 = f(inp)
+    # From options; P('x', 'y'), P('y'), shard_like chooses the better option.
+    self.assertEqual(out1.sharding, s)
+    self.assertEqual(out2.sharding, s)
+
+  def test_grad(self):
+    mesh = jtu.create_global_mesh((4,), ('x',))
+    np_inp = np.arange(8.)
+    s = NamedSharding(mesh, P('x'))
+    inp = jax.device_put(np_inp, s)
+
+    def _cb(s):
+      self.assertFalse(s.is_fully_replicated)
+      self.assertLen(s.device_set, mesh.size)
+      self.assertEqual(s.shard_shape(np_inp.shape), (2,))
+
+    def f(x):
+      y = jnp.arange(8.)
+      x_, y_ = shard_alike(x, y)
+      jax.debug.inspect_array_sharding(y_, callback=_cb)
+      z = x_ + y_
+      return jnp.sum(z)
+
+    jax.grad(f)(inp)  # doesn't crash
+    jax.grad(jax.jit(f))(inp)  # doesn't crash
+
+  def test_shard_input_as_output(self):
+    mesh = jtu.create_global_mesh((4,), ('x',))
+    np_inp = np.arange(8.)
+    s = NamedSharding(mesh, P('x'))
+
+    @jax.jit
+    def f(x):
+      y = jax.lax.with_sharding_constraint(x, s)
+      z = y * 2
+      return shard_alike(x, z)
+
+    with jtu.count_pjit_cpp_cache_miss() as count:
+      f(np_inp)
+      out1, out2 = f(np_inp)
+    self.assertEqual(count[0], 1)
+    self.assertTrue(s.is_equivalent_to(out1.sharding, np_inp.ndim))
+    self.assertTrue(s.is_equivalent_to(out2.sharding, np_inp.ndim))
+
+    @jax.jit
+    def g(x):
+      z = x * 2
+      return shard_alike(x, z)
+    arr = jax.device_put(np_inp, s)
+    with jtu.count_pjit_cpp_cache_miss() as count:
+      g(arr)
+      out3, out4 = g(arr)
+    self.assertEqual(count[0], 1)
+    self.assertEqual(out3.sharding, s)
+    self.assertEqual(out4.sharding, s)
+
+  def test_shard_alike_inputs(self):
+    mesh = jtu.create_global_mesh((2,), ('x',))
+    np_inp = np.arange(8.)
+    s = NamedSharding(mesh, P('x'))
+    rep_s = NamedSharding(mesh, P())
+    arr = jax.device_put(np_inp, s)
+    arr2 = jax.device_put(np_inp, rep_s)
+
+    def f(x, y):
+      return shard_alike(x, y)
+
+    eager_out1, eager_out2 = f(arr, arr2)
+    self.assertEqual(eager_out1.sharding, s)
+    self.assertEqual(eager_out2.sharding, s)
+
+    out1, out2 = jax.jit(f)(arr, arr2)
+    self.assertEqual(out1.sharding, s)
+    self.assertEqual(out2.sharding, s)
+
+
+if __name__ == '__main__':
+  absltest.main(testLoader=jtu.JaxTestLoader())
